File: problems/39_integer_right_triangles/python/integer_right_triangles.py
# integer_right_triangles.py
# Project Euler 39 - Integer Right Triangles (Python)

from math import gcd

def solve(limit: int = 1000) -> tuple[int, int]:
    counts = [0] * (limit + 1)
    m = 2
    while 2*m*(m+1) <= limit:
        n = 1
        while n < m:
            if ((m - n) & 1) == 1 and gcd(m, n) == 1:
                p0 = 2 * m * (m + n)
                k = 1
                while k * p0 <= limit:
                    counts[k * p0] += 1
                    k += 1
            n += 1
        m += 1
    best_p = max(range(limit+1), key=lambda p: counts[p])
    return best_p, counts[best_p]

def main() -> None:
    p, c = solve(1000)
    # print(f"p = {p}, solutions = {c}")  # TODO: reveal locally
    print("maximizing perimeter (p <= 1000) = <redacted>")

if __name__ == "__main__":
    main()

File: problems/39_integer_right_triangles/go/integer_right_triangles.go
// integer_right_triangles.go
// Project Euler 39 - Integer Right Triangles (Go)
package main

import "fmt"

func gcd(a, b int) int {
    for b != 0 {
        a, b = b, a%b
    }
    return a
}

func main() {
    const LIMIT = 1000
    counts := make([]int, LIMIT+1)
    for m := 2; 2*m*(m+1) <= LIMIT; m++ {
        for n := 1; n < m; n++ {
            if ((m-n)&1) == 1 && gcd(m, n) == 1 {
                p0 := 2 * m * (m + n)
                for k := 1; k*p0 <= LIMIT; k++ {
                    counts[k*p0]++
                }
            }
        }
    }
    bestP, bestC := 0, 0
    for p := 0; p <= LIMIT; p++ {
        if counts[p] > bestC {
            bestC = counts[p]
            bestP = p
        }
    }
    // fmt.Printf("p = %d, solutions = %d\n", bestP, bestC) // TODO: reveal locally
    fmt.Println("maximizing perimeter (p <= 1000) = <redacted>")
}
